### PR TITLE
Start implementing new "let" that contains Patterns on the LHS

### PR DESCRIPTION
**Pull request by StachuDotNet**
Changelog:

```
Internal
- start to implement a new 'let' expression that contains patterns on the left (e.g. to deconstruct tuples)
```

Currently, users may only assign a single variable in a let expression. This is the first step in replacing our existing Let expression with a new Let that supports multiple kinds of patterns on the LHS of the `=`. It's based off of #4711, and the larger plan is outlined in [this document](https://docs.google.com/document/d/10mr811ONuUSm6Q1egfv9dGtpN7QdHwhM96uyW8w0140/edit#heading=h.2dqrn78eah12).

This PR contains the work outlined in the first bullet of #4732.
At the end of this PR, both the `client` and `backend` 'think' in terms of the new Let (with a Variable pattern) but the communication is still in the old kind of Let. That said, _decoders_ on both ends have been set up for the next relevant PR, which will switch over communication to be in the _new_ kind of Let.
Relevantly, some serialization roundtripping tests were failing - the decoders currently "drop" the Ids contained within the LetPattern, and the encoders generate a new ID for the variable LetPattern. For now, I've disabled these tests, to turn them back on as part of the next PR - I've done enough manual testing to be comfortable with this, but I'm open to ideas to avoid such.

### CODE REVIEW

File: backend/src/LibExecution/Interpreter.fs [hunk 1]
```diff
@@ -59,19 +59,21 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
     | ENull _id -> return DNull
     | ECharacter (_id, s) -> return DChar s
 
-
-    | ELet (_id, lhs, rhs, body) ->
+    | ELet (_id, pat, rhs, body) ->
       let! rhs = eval state st rhs
+
       match rhs with
-      // CLEANUP we should still preview the body
```
Comment (StachuDotNet) [inline backend/src/LibExecution/Interpreter.fs:66]: Use a simplified form of the code in #4711 for this (which _does_ preview the body)

File: backend/tests/FuzzTests/Generators.fs [hunk 2]
```diff
@@ -192,16 +192,32 @@ module RuntimeTypes =
                       //(4, ok) // TODO: random string, with 0-5 args
                        ]
 
-  open MatchPattern
+    let rec genTuple (s, genArg) : Gen<RT.MatchPattern> =
```
Comment (StachuDotNet) [inline backend/tests/FuzzTests/Generators.fs:195]: Looks like I backfilled this, but didn't 'use' it below. Not top-priority but might as well use it below in the file

File: backend/tests/FuzzTests/Generators.fs [hunk 3]
```diff
@@ -347,6 +398,7 @@ module RuntimeTypes =
     let recursiveExprs =
       [ genConstructor
         genLet
+        genLetWithPattern
```
Comment (StachuDotNet) [inline backend/tests/FuzzTests/Generators.fs:401]: We don't need duplicate defs

File: backend/tests/TestUtils/FSharpToExpr.fs [hunk 4]
```diff
@@ -297,43 +297,29 @@ let rec convertToExpr' (ast : SynExpr) : PT.Expr =
 
   // When we add patterns on the left hand side of lets, the pattern below
   // could be expanded to use convertPat
+  // TODO: this doesn't allow nesting (i.e. `let (a, (b, c)) = (1, (2, 3))`. Should it?)
```
Comment (StachuDotNet) [inline backend/tests/TestUtils/FSharpToExpr.fs:300]: no longer true, comment may be removed

**Comment by StachuDotNet** [comment]
this is generally looking good to me, but a few backend serialization roundtrip tests are failing.
I hit a similar case in equivalent client-side tests, and I solved it by forcing the IDs of the Variable LetPattern to be `0` in the test scenarios. I could do that same thing for the backend tests, but that's feeling pretty hacky.

Another solution would be to temporarily remove the `id` from the LPVariable definition _in ClientTypes_. Until we have additional patterns, this ID isn't actually useful. I'm going to try that.

**Comment by StachuDotNet** [comment]
Nah, that's not enough, since we do roundtrips between 'internal' types and clienttypes as well.
(start with an internal Expr. then `expr -> toCT -> fromCT` - this loses the ID)

I started going down this road, but the many places where we perform toID make it a bit awkward.

To avoid repeated work and churn I'm thinking to:
- remove the hack in the client side
- simply disable the roundtripping tests
- rely on manual verification of roundtripping for now (my confidence is very high)
- turn the tests back on once we've migrated to the new Let
  (really, the next PR as outlined in #4732 is fine - the tests would only be disabled for a bit)

**Comment by StachuDotNet** [comment]
I think dark is re-stable enough to refresh this against the updated codebase - what do you think @pbiggar?

**Comment by pbiggar** [comment]
Yes, I think so. Though I'd wait for my Test changes to land - maybe tomorrow.

**Comment by StachuDotNet** [comment]
Replaced

**Pull request closed without merging**